### DISCUSSION TITLE
Support for an incoming-only keypair / fix multi-user security in OSS

### DISCUSSION THREAD

**Discussion by pylorak** [rustdesk/rustdesk#9213] [General]
TLDR; Have the server generate not one, but two keypairs. The second keypair can only be used for incoming-only connections.

Details:

Here is a problematic, but probably common scenario: User A has a self- or private-hosting setup for **OSS** RustDesk server. User A uses RustDesk primarily to access his own devices, but occasionally, he needs a remote connection to a friend or maybe even an untrusted party (someone met on the internet, Steam friend, chat friend etc.). So user A shares his server address and public key to User B to dial into B's computer. 

Now the server address and pubkey are the wild. User A has completely lost control over who else can access his server. User B can install RustDesk on any of his devices and use A's server without his knowledge. Worse, if user B shares the settings with user C and D, and they share the same settings with even more users, soon "half the world" is using user A's private infrastructure.
User A of course can generate a new keypair. However, this means he has to reconfigure all his devices, and also has to do so every time he connects to another user's device.

**Note this also means that OSS RustDesk is currently principally insecure in a multi-user setup (as soon as you share your server address and pubkey, even if just momentarily, you have lost control over who and how many can use your infrastructure, unless you reconfigure the server and all your own clients).** 

Optimally such a scenario would be handled with some basic user-management, but I see how this would eat into RustDesk's revenue stream.

So instead, the suggested solution is to have OSS RustDesk generate and support a secondary keypair. When a client connects to the server using the secondary keypair, it can only accept incoming connections, but cannot initiate new ones itself. This allows to easily secure A's infrastructure without having to support the web-console or any kind of user management. In the above example, if user A shares only the secondary pubkey but not the primary pubkey, his friends and his friend's friends cannot use A's server unauthorized anymore.

**Comment by basilgello** [comment]
Interesting usecase but let me understand it properly:

1. Knowing hbbr/hbbs server address and pubkey can lead to abuse of that relay server bandwidth (or DDoS it!) but it has nothing to do with access to end user device as permanent access password, one-time password or confirmation by local user is needed to initiate the session.
2. Knowing API key for server grants access to address book but user A should not share it in order to let user B in.

**Reply by pylorak** [reply]
> Interesting usecase but let me understand it properly:
> 
>     1. Knowing hbbr/hbbs server address and pubkey can lead to abuse of that relay server bandwidth (or DDoS it!) but it has nothing to do with access to end user device as permanent access password, one-time password or confirmation by local user is needed to initiate the session.

Correct. The concern here is not that users B/C/D... could get access to user A's devices. The concern is that they get to use user A's server resources, and user A has no control over how many or who. Currently, anybody who knows the server's address and public key can share these and start new connections to 3rd parties without user A's consent. This means basically that user A must not (should not) share the address and pubkey with anybody, which also means OSS RustDesk is only suitable/secure if you use it exclusively on your own devices.

> 
>     2. Knowing API key for server grants access to address book but user A should not share it in order to let user B in.

I thought the address book is not usable in the OSS version because it requires a login. But if I overlooked something and simply knowing the public key grants access to the address book, then it would be a serious privacy concern. It means if I want to connect to anybody, they could see my address book. A big no-no. If this is really the case, then in the proposed solution, the secondary public key must also prohibit access to the address book. The idea is that the with the secondary public key, a user should only be able to receive incoming connections and have otherwise zero capabilities/privileges.
